crypto/s2n_hash.c

```diff
@@ -169,13 +173,270 @@ int s2n_hash_digest(struct s2n_hash_state *state, void *out, uint32_t size)
     return 0;
 }
 
-int s2n_hash_reset(struct s2n_hash_state *state)
+static int s2n_low_level_hash_copy(struct s2n_hash_state *to, struct s2n_hash_state *from)
 {
-    return s2n_hash_init(state, state->alg);
+    memcpy_check(to, from, sizeof(struct s2n_hash_state));
+    return 0;
 }
 
-int s2n_hash_copy(struct s2n_hash_state *to, struct s2n_hash_state *from)
+static int s2n_low_level_hash_reset(struct s2n_hash_state *state)
 {
-    memcpy_check(to, from, sizeof(struct s2n_hash_state));
+    return s2n_low_level_hash_init(state, state->alg);
+}
+
+static int s2n_low_level_hash_free(struct s2n_hash_state *state)
+{
+    return 0;
+}
+
+static int s2n_evp_hash_new(struct s2n_hash_state *state)
+{
+    notnull_check(state->digest.evp.primary.ctx = S2N_EVP_MD_CTX_NEW());
+    notnull_check(state->digest.evp.md5_secondary.ctx = S2N_EVP_MD_CTX_NEW());
```

Comment:
Yes, it's only used for md5_sha1. This can't be much cleaner as we want to preserve s2n_hash_state's ability to be reset/init'ed to a different hash algorithm without additional allocation required. This allows us to pre-allocate all memory needed for all s2n_hash_algorthms.